File: Algorithms_And_Abstract_Data_Types/pa2/List.c
/*
Adel Danandeh 
CMPS 101 - Project #2

*/

#include<stdio.h>
#include<stdlib.h>
#include"List.h"

List newList(void){
	List newList = (List) malloc (sizeof(ListObject));

	newList->front = NULL;
	newList->back = NULL;
	newList->cursor = NULL;
	newList->length = 0;

	return newList;
}

int length(List L){

	return L->length;

}

int getIndex(List L) {

	int index = -1;
	Node temp = L->cursor;

	while (temp != NULL) {

		temp = temp->backward;
		index++;
	}

	return index;

}

int front(List L){

	if(length(L) > 0){

		return L->front->data;
	}
	return -1;
}

int back(List L){

	if(length(L) > 0){
		return L->back->data;
	}

	return -1;
}

int getElement(List L){

	if(length(L) > 0, getIndex(L) >= 0){
		return L->cursor->data;
	}

	return -1;
}

int equals(List A, List B){

	Node tempNode1 = A->front;
	Node tempNode2 = B->front;

	while (tempNode1 != NULL && tempNode2 != NULL) {

		if (tempNode1->data != tempNode2->data) {

			return 0;
		}

		tempNode1 = tempNode1->forward;
		tempNode2 = tempNode2->forward;
	}

	return 1;
}

void moveTo(List L, int i){

	int j = 0;
	L->cursor = L->front;

	if (length(L) > 0) {
		while (i > j) {

			L->cursor = L->cursor->forward;
			j++;
		}
	} else

		L->cursor = NULL;

	return;
}

void movePrev(List L){

	if (getIndex(L) <= length(L) - 1) {
		L->cursor = L->cursor->backward;
	}

	else if (getIndex(L) == 0) {

		L->cursor = NULL;
	}
	return;
}
void moveNext(List L){

	if (getIndex(L) < length(L) - 1) {

		L->cursor = L->cursor->forward;
	}

	else if (getIndex(L) == length(L) - 1) {
		L->cursor = NULL;
	}
	return;
}

void prepend(List L, int data){

	Node newNode = (Node) malloc (sizeof(NodeObject));

	if (length(L) == 0) {

		L->front = newNode;
		L->back = newNode;
		newNode->backward = NULL; 
		newNode->forward = NULL;
	}

	else {


		L->front->backward = newNode;
		newNode->forward = L->front;
		newNode->backward = NULL;
		L->front = newNode;
	}
	newNode->data = data;
	L->length++;

	return;
}

void append(List L, int data){

	Node newNode = (Node) malloc (sizeof(NodeObject));

	if (length(L) == 0) {

		L->front = newNode;
		L->back = newNode;
		newNode->backward = NULL; 
		newNode->forward = NULL;

	}

	else {

		L->back->forward = newNode;
		newNode->backward = L->back;
		L->back = newNode;
		newNode->forward = NULL;

	}
	newNode->data = data;
	L->length++;

	return;
}

void insertBefore(List L, int data){

	if (getIndex(L) >= 0 && length(L) > 0) {

		Node newNode = (Node) malloc (sizeof(NodeObject));
		newNode->data = data;

		if (L->cursor->backward == NULL) {

			L->cursor->backward = newNode;
			newNode->forward = L->cursor;
			L->front = newNode;
			newNode->backward = NULL;
		}

		else {

			newNode->forward = L->cursor;
			newNode->backward = L->cursor->backward;
			L->cursor->backward->forward = newNode;
			L->cursor->backward = newNode;
		}
	}

	L->length++;

	return;
}

void insertAfter(List L, int data){

	if (getIndex(L) >= 0 && length(L) > 0) {

		Node newNode = (Node) malloc (sizeof(NodeObject));
		newNode->data = data;

		if (L->cursor->forward == NULL) {

			L->cursor->forward = newNode;
			newNode->backward = L->cursor;
			newNode->forward = NULL;
			L->back = newNode;
		}

		else {

			newNode->backward = L->cursor;
			newNode->forward = L->cursor->forward;
			L->cursor->forward->backward = newNode;
			L->cursor->forward = newNode;

		}

	}

	L->length++;

	return;
}

void deleteFront(List L){

	if (length(L) > 0) {

		if (L->front->forward == NULL) {

			free(front);
			L->front = NULL;
			L->back = NULL;	
		}

		else {

			L->front = L->front->forward;
			free(L->front->backward);
			L->front->backward = NULL;
		}

	}

	L->length--;

	return;
}

void deleteBack(List L){

	if (length(L) > 0) {

		if (L->back->backward == NULL) {

			free(free);
			L->back = NULL;
			L->front = NULL;
		}

		else {

			L->back = L->back->backward;
			free(L->back->backward);
			L->back->backward = NULL;
		}
	}
	L->length--;

	return;
}

void delete(List L){

	if (getIndex(L) >= 0 && length(L) > 0) {

		if (L->length == 1) {

			L->cursor->forward = NULL;
			L->cursor->backward = NULL;

		}

		else if(getIndex(L) == 0){

			L->cursor->forward = NULL;
			L->cursor->backward = NULL;
			L->front = NULL;
			L->back = NULL;
		}

		else if (getIndex(L) > 0) {

			L->cursor->forward->backward = L->cursor->backward;
			L->cursor->backward->forward = L->cursor->forward;
			L->cursor->forward = NULL;
			L->cursor->backward = NULL;

		}

		else if (getIndex(L) == L->length-1) {

			L->cursor->backward->forward = NULL;
			L->cursor->backward = NULL;
			L->cursor->forward = NULL;
		}
	}

	free(L->cursor);
	L->length--;
	return;
}

List copyList(List L) {

	Node temp = L->front;
	List newL;
	int i;
	newL = newList();

	for (i = 0; i < L->length-1; i++) {

		append(newL, temp->data);
		temp = temp->forward;
	}

	return newL;
}

void printList(FILE* out, List L){
	
	Node temp = L->front;

	while(temp != NULL){

		fprintf(out, "%d " , temp->data);
		temp = temp->forward;
	}
}

void clear(List L){

	Node temp = L->front;
	Node temp1;

	while(temp != NULL){

		temp1 = temp;
		temp = temp->forward;
		free(temp1);
	}


	L->front = NULL;
	L->back = NULL;
	L->length = 0;
}

void freeList(List* L){

	clear(*L);
	free(*L);
}

File: Algorithms_And_Abstract_Data_Types/pa3/List.java
/* 
 Adel Danandeh
 Assignment #3

 List.java is implementation of a doubly link list structure which has certain methods defined in it. 

 */

public class List {

	private class Node {

		private Node forward;
		private Node backward;
		private Object data;

		Node() {

			forward = null;
			backward = null;
			this.data = null;
		}
	}

	private Node front;
	private Node back;
	private int length;
	private Node cursor;

	List() {

		front = null;
		back = null;
		cursor = null;
		this.length = 0;
	}

	// The function length does take any arguments
	// This function returns the length of the list
	public int length() {

		return length;
	}

	// The function getIndex does not take any arguments.
	// Returns the index of the cursor element in this list, or
	// returns -1 if the cursor element is undefined.
	public int getIndex() {
		int index = -1;
		Node temp;
		temp = cursor;

		while (temp != null) {

			temp = temp.backward;
			index++;
		}

		return index;
	}

	// The function front does not take any arguments.
	// Pre: length()>0
	Object front() {

		if (length() > 0) {
			return front.data;
		}

		return null;
	}

	// Function back does not take arguments
	// Pre: length()>0
	Object back() {

		if (length() > 0) {

			return back.data;
		}

		return null;
	}

	// The function getElement does not take any arguments.
	// Pre: length()>0, getIndex()>=0
	Object getElement() {

		if (length() > 0 && getIndex() >= 0)
			return cursor.data;

		return null;
	}

	// The function equals takes a list as an argument.
	// Returns true if this List and L are the same integer
	// sequence. The cursor is ignored in both lists.
	public boolean equals(Object x) {

		Node tempNode1 = front;
		Node tempNode2 = ((List) x).front;

		while (tempNode1 != null && tempNode2 != null) {

			if (tempNode1.data != tempNode2.data) {

				return false;
			}

			tempNode1 = tempNode1.forward;
			tempNode2 = tempNode2.forward;

		}

		return true;
	}

	// Function clear does not take any arguments.
	// Re-sets this List to the empty state.
	void clear() {

		length = 0;
		front = back = null;
		cursor = null;
	}

	// Function moveTo takes an integer as an argument.
	// If 0<=i<=length()-1, moves the cursor to the element
	// at index i, otherwise the cursor becomes undefined.
	void moveTo(int i) {

		int counter = 0;
		if (i >= 0 && i <= length() - 1) {
			cursor = front;
			while (counter != i) {
				cursor = cursor.forward;
				counter++;
			}
		} else
			cursor = null;
	}

	// Function movePrev does not take any arguments
	// If 0<getIndex()<=length()-1, moves the cursor one step toward the
	// front of the list. If getIndex()==0, cursor becomes undefined.
	// If getIndex()==-1, cursor remains undefined.
	void movePrev() {

		if (getIndex() > 0 && getIndex() <= length() - 1) {
			cursor = cursor.backward; // front of the list
		} else {
			cursor = null;
		}
	}

	// Function moveNext does not take any arguments
	// If 0<=getIndex()<length()-1, moves the cursor one step toward the
	// back of the list. If getIndex()==length()-1, cursor becomes
	// undefined. If index==-1, cursor remains undefined
	void moveNext() {

		if (0 <= getIndex() && getIndex() < length() - 1) {
			cursor = cursor.forward; // back of the list
		} else {
			cursor = null;
		}

	}

	// Function prepend takes an object as an argument
	// Inserts new element before front element in this List.
	void prepend(Object data) {

		Node newNode = new Node();

		if (length() == 0) {

			back = newNode;
		}

		else {

			front.backward = newNode;
			newNode.forward = front;
		}
		front = newNode;
		newNode.data = data;
		length++;
	}

	// Function append takes an object as an argument
	// Inserts new element after back element in this List.
	void append(Object data) {

		Node newNode = new Node();
		newNode.data = data;
		if (length() == 0)
			front = newNode;
		else {
			back.forward = newNode;
			newNode.backward = back;
		}
		back = newNode;
		length++;
	}

	// Function inserBefore takes an object data.
	// Inserts new element before cursor element in this
	// Pre: length()>0, getIndex()>=0
	void insertBefore(Object data) {

		if (getIndex() >= 0 && length() > 0) {

			Node newNode = new Node();
			newNode.data = data;

			if (getIndex() == 0) {

				cursor.backward = newNode;
				newNode.forward = cursor;
				front = newNode;
			}

			else {

				newNode.forward = cursor;
				newNode.backward = cursor.backward;
				cursor.backward.forward = newNode;
				cursor.backward = newNode;
			}
		}
		length++;
	}

	// Function insertAfter takes an object as argument.
	// Inserts new element after cursor element in this
	// Pre: length()>0, getIndex()>=0
	void insertAfter(Object data) {

		if (getIndex() >= 0 && length() > 0) {

			Node newNode = new Node();
			newNode.data = data;

			if (getIndex() == 0) {

				cursor.forward = newNode;
				newNode.backward = cursor;

			}

			else if (cursor.forward == null) {

				cursor.forward = newNode;
				newNode.backward = cursor;
			}

			else {

				newNode.backward = cursor;
				newNode.forward = cursor.forward;
				cursor.forward.backward = newNode;
				cursor.forward = newNode;

			}

		}

		length++;
	}

	// Function deleteFront does not take any arguments
	// Deletes the front element in this List. Pre: length()>0
	void deleteFront() {

		if (length() >= 0) {
			if (length() == 1) {
				front = null;
				back = null;
			} else {
				front = front.forward;
				front.backward.forward = null;
				front.backward = null;
			}
			length--;
		}
	}

	// Function deleteBack does not take any arguments.
	// Deletes the back element in this List. Pre: length()>0
	void deleteBack() {

		if (length() >= 0) {
			if (length() == 1) {
				front = null;
				back = null;
			} else {
				back = back.backward;
				back.forward.backward = null;
				back.forward = null;
			}
			length--;
		}
	}

	// Function delete does not take any arguments
	// Deletes cursor element in this List. Cursor is undefined after this
	// operation. Pre: length()>0, getIndex()>=0
	void delete() {
		if (length() > 0 && getIndex() >= 0) {
			if (getIndex() == 0) {
				if (length() > 1)
					cursor.forward.backward = null;
				else
					back = null;
				front = cursor.forward;
			} else if (getIndex() == length() - 1) {
				cursor.backward.forward = null;
				back = cursor.backward;
			} else {
				cursor.backward.forward = cursor.forward;
				cursor.forward.backward = cursor.backward;
			}
			cursor = null;
			length--;
		}
	}

	public String toString() {

		String str = " ";
		Node temp = front;

		for (int i = 0; i < length; i++) {

			str += temp.data.toString() + " ";
			temp = temp.forward;
		}

		return str;
	}
}
